### PR TITLE
Write documentation for @Proxy

### PR DESCRIPTION
**Pull request by lidaamber**
Add general proxy description and a few examples for its annotations
referencing #17

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+```java
+// Creates singleton proxy with default interface name defaultInterface
+@Proxy(interfaceName = "defaultInterface", singleton = true)
+public interface MyProxy {
+  // proxy methods
+}
+```
+
+For now, you can declare proxy wrappers for sending `call` and `event` packets.
+Proxy will be extended for other packets handling.
```
Comment (lundibundi) [inline README.md]: Not sure about this line, there may be new features but necessarily connected to that.
Anyway s/for other packets handling/to handle other packets.

File: README.md [hunk 2]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+
+##### Call
+
+To create wrapper for `call` packet, use `@Call` annotation. Set the method
```
Comment (lundibundi) [inline README.md]: s/packet/method

File: README.md [hunk 3]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+name as an annotation parameter (the default interface will be used), or
```
Comment (lundibundi) [inline README.md]: It is not necessary to set method name, by default java method name will be used. We should state that.

File: README.md [hunk 4]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+interface name and method name. Specify the arguments and callback handler
+(if they are required) as your method parameters. See the examples:
```
Comment (lundibundi) [inline README.md]: s/required/needed

File: README.md [hunk 5]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+
+```java
+
+@Proxy(interfaceName = "defaultInterface")
+public interface MyProxy {
```
Comment (lundibundi) [inline README.md]: We should add example with just @Call.

File: README.md [hunk 6]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+  // Makes a call with specified interfaceName and methodName, sets param and
+  // otherParam as call arguments and uses hander on callback
```
Comment (lundibundi) [inline README.md]: uses 'hander' to handle method callback

File: README.md [hunk 7]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+  @Call({"interfaceName", "methodName"})
+  void callInterfaceMethod(String param, int otherParam, ManualHandler handler);
+
+  // Makes a call with default interface name and methodName without arguments
+  // and uses hander on callback
```
Comment (lundibundi) [inline README.md]: ditto.

File: README.md [hunk 8]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+  // Makes a call with default interface name and otherMethodName without
+  // arguments and without callback handler
+  @Call("otherMethodName")
+  void callNoParametersMethod();
+}
+```
+
+##### Event
+
+To create wrapper for `event` packet, use `@Event` annotation. Set the method
```
Comment (lundibundi) [inline README.md]: s/packet/method

File: README.md [hunk 9]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+name as an annotation parameter (the default interface will be used), or
```
Comment (lundibundi) [inline README.md]: ditto about default method name.

File: README.md [hunk 10]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+interface name and method name. Specify the arguments (if they are required) as
```
Comment (lundibundi) [inline README.md]: s/required/needed

File: README.md [hunk 11]
```diff
@@ -442,3 +442,91 @@ receiver.addHandler(new ExampleReceiver() {
...
+your method parameters. See the examples:
+
+```java
+@Proxy(interfaceName = "defaultInterface")
+public interface MyProxy {
+  // Sends event with specified interfaceName and eventName without parameters
```
Comment (lundibundi) [inline README.md]: ditto about default method name example.

**Comment by lundibundi** [comment]
Also, will you add examples of Handler usage in connection with proxy or I may add it later? Same goes for more usage examples of Proxy?

**Comment by lidaamber** [comment]
@lundibundi yes, I can add more examples. Do you mean adding handlers for calls and events?

**Comment by lundibundi** [comment]
@lidaamber yeah, first part is about examples of @Handlers with @Proxy addEventHandler and setCallHandler. The second part is usage of generated proxy (making calls, sending events).

File: README.md [hunk 12]
```diff
@@ -442,3 +441,242 @@ receiver.addHandler(new ExampleReceiver() {
 });
 connection.call("interfaceName", "methodName", args, receiver);
 ```
+
+#### @Proxy
+
+`@Proxy` annotation allows you to auto-generate interface to the remote
+server that will look as if you are calling local java methods easily. To create
```
Comment (lundibundi) [inline README.md]: `@Proxy` annotation allows you to auto-generate interface to the remote
 +server that will look as if you are calling local java methods easily
->
`@Proxy` annotation allows you to easily auto-generate interface to the remote
 +server that will look as if you are calling local java methods

File: README.md (hunk 1 above)
Comment (lundibundi) [inline README.md]: @lidaamber We should probably remove this line as it's not at all informative.

File: README.md [hunk 13]
```diff
@@ -442,3 +441,242 @@ receiver.addHandler(new ExampleReceiver() {
...
+  void callDefaultInterfaceMethod(ManualHandler handler);
+
+  // Makes a call with default interface name and method name "otherMethodName"
+  // without arguments and without callback handler
+  @Call("otherMethodName")
+  void callNoParametersMethod();
+
+  // Makes a call with default interface name and method name "joinChat"
+  // without arguments and uses "handler" to handle method callback
+  @Call()
```
Comment (lundibundi) [inline README.md]: The parenthesis are unnecessary.

File: README.md [hunk 14]
```diff
@@ -442,3 +441,242 @@ receiver.addHandler(new ExampleReceiver() {
...
+  void sendInterfaceEvent();
+
+  // Sends event with default interface name and event name "eventName" with
+  // specified parameters
+  @Event("eventName")
+  void sendEvent(String param);
+
+  // Sends event with default interface name and event name
+  // "onSomethingHappened" without parameters
+  @Event()
```
Comment (lundibundi) [inline README.md]: ditto.

File: README.md [hunk 15]
```diff
@@ -442,3 +441,242 @@ receiver.addHandler(new ExampleReceiver() {
...
+proxy.addEventHandler("interfaceName", "eventName", new ManualHandler() {
+    @Override
+    public void handle(JSObject jsObject) {
+        // handle object
+    }
+});
+
+// adds executable handler for "eventName" of "interfaceName" (see the example
+// of executable handler)
+proxy.addEventHandler("interfaceName", "eventName",
```
Comment (lundibundi) [inline README.md]: I think it's better to add example of handler with ExecutableHandler as its base, not just ExecutableHandler itself.

File: README.md [hunk 16]
```diff
@@ -494,8 +493,8 @@ public interface MyProxy {
   @Call("otherMethodName")
   void callNoParametersMethod();
 
-  // Makes a call with default interface name and method name "joinChat"
-  // without arguments and uses "handler" to handle method callback
+  // Makes a call with default interface name and method name joinChat
+  // without arguments and uses handler to handle method callback
   @Call()
```
Comment (lundibundi) [inline README.md]: parenthesis after Call =). Please put the quotes back.

File: README.md [hunk 17]
```diff
@@ -524,7 +523,7 @@ public interface MyProxy {
   void sendEvent(String param);
 
   // Sends event with default interface name and event name
-  // "onSomethingHappened" without parameters
+  // onSomethingHappened without parameters
   @Event()
```
Comment (lundibundi) [inline README.md]: ditto

File: README.md [hunk 18]
```diff
@@ -680,3 +679,28 @@ proxy.addEventHandler("interfaceName", "eventName", new JSTPEventHandler() {
     }
 });
 ```
+
+If we modify the annotation in such a way:
```
Comment (lundibundi) [inline README.md]: maybe better in such a way -> this way / like this?

**Comment by lundibundi** [comment]
Closed via https://github.com/metarhia/jstp-java/commit/807e82075438370992ad2f4abc215ecee4cc289f. :tada:

**Pull request closed without merging**